File: android/src/main/kotlin/com/cloudwebrtc/webrtc/model/MediaType.kt
package com.cloudwebrtc.webrtc.model

import org.webrtc.MediaStreamTrack.MediaType as WMediaType

/**
 * [org.webrtc.MediaStreamTrack.MediaType] representation.
 *
 * @property value [Int] representation of this enum which will be expected on the Flutter side.
 */
enum class MediaType(val value: Int) {
  AUDIO(0),
  VIDEO(1);

  companion object {
    /**
     * Tries to create a [MediaType] based on the provided [Int].
     *
     * @param value [Int] value from which [MediaType] will be created.
     *
     * @return [MediaType] based on the provided [Int].
     */
    fun fromInt(value: Int) = values().first { it.value == value }
  }

  /**
   * Converts this [MediaType] into an [org.webrtc.MediaStreamTrack.MediaType].
   *
   * @return [org.webrtc.MediaStreamTrack.MediaType] based on this [MediaType].
   */
  fun intoWebRtc(): WMediaType {
    return when (this) {
      AUDIO -> WMediaType.MEDIA_TYPE_AUDIO
      VIDEO -> WMediaType.MEDIA_TYPE_VIDEO
    }
  }
}

File: ios/Classes/model/MediaType.swift
import WebRTC

/// Representation of an `RTCRtpMediaType`.
enum MediaType: Int {
  /// Audio media.
  case audio

  /// Video media.
  case video

  /// Converts the provided `RTCRtpMediaType` into a `MediaType`.
  static func fromWebRtc(kind: RTCRtpMediaType) -> MediaType? {
    switch kind {
    case .audio:
      return MediaType.audio
    case .video:
      return MediaType.video
    default:
      return nil
    }
  }

  /// Creates this `RTCRTPMediaType` from `String`.
  static func fromString(kind: String) -> MediaType? {
    switch kind {
    case "audio":
      return MediaType.audio
    case "video":
      return MediaType.video
    default:
      return nil
    }
  }

  /// Converts this `MediaType` into an `RTCRTPMediaType`.
  func intoWebRtc() -> RTCRtpMediaType {
    switch self {
    case .audio:
      return RTCRtpMediaType.audio
    case .video:
      return RTCRtpMediaType.video
    }
  }

  /// Converts this `MediaType` into a `String`.
  func toString() -> String {
    switch self {
    case .audio:
      return "audio"
    case .video:
      return "video"
    }
  }
}
